File: frontend/src/lib/media-fields-catalog.ts
/**
 * Media fields visibility and categorization based on library type (Frontend).
 * Used to determine which columns/panels/filters should be shown for each library type.
 */

import type { LibraryType } from "./api";

export type FieldCategory = "video_only" | "audio_only" | "music_only" | "audiobook_only" | "shared";

// Map of fields to their categories
export const FIELD_CATEGORIES: Record<string, FieldCategory> = {
  // Video-only fields
  video_codec: "video_only",
  resolution: "video_only",
  hdr_type: "video_only",
  bitrate: "video_only",

  // Shared fields
  file: "shared",
  size: "shared",
  container: "shared",
  duration: "shared",
  quality_score: "shared",

  // Audio fields (appear in all types that support audio)
  audio_codec: "audio_only",
  audio_spatial_profile: "audio_only",
  audio_language: "audio_only",
  audio_bitrate: "audio_only",
  bit_depth: "audio_only",

  // Music-only fields
  audio_title: "music_only",
  audio_artist: "music_only",
  audio_album: "music_only",
  audio_album_artist: "music_only",
  audio_genre: "music_only",
  audio_date: "music_only",
  audio_disc: "music_only",
  audio_composer: "music_only",
  audio_channels: "audio_only",
  sample_rate: "audio_only",
  track_number: "music_only",
  bit_rate_mode: "audio_only",
  has_embedded_cover: "music_only",
  chapter_count: "audiobook_only",
  chapter_titles: "audiobook_only",
  audiobook_narrator: "audiobook_only",
  audiobook_author: "audiobook_only",
  audiobook_publisher: "audiobook_only",
  audiobook_series: "audiobook_only",
  audiobook_series_part: "audiobook_only",
  audiobook_language: "audiobook_only",
  audiobook_abridged: "audiobook_only",
  audiobook_asin: "audiobook_only",
  audiobook_isbn: "audiobook_only",
};

/**
 * Determine if a field should be visible for a given library type.
 */
export function shouldShowField(fieldKey: string, libraryType: LibraryType): boolean {
  const category = FIELD_CATEGORIES[fieldKey] ?? "shared";

  switch (category) {
    case "video_only":
      return libraryType !== "music" && libraryType !== "audiobooks";
    case "music_only":
      return libraryType !== "movies" && libraryType !== "series";
    case "audiobook_only":
      return libraryType === "audiobooks";
    case "audio_only":
      return true; // Audio is in all types
    case "shared":
      return true; // Shared fields always visible
    default:
      return true;
  }
}

/**
 * Get all visible field keys for a library type.
 */
export function getVisibleFieldsForType(libraryType: LibraryType): Set<string> {
  return new Set(
    Object.entries(FIELD_CATEGORIES)
      .filter(([fieldKey]) => shouldShowField(fieldKey, libraryType))
      .map(([fieldKey]) => fieldKey)
  );
}

File: backend/app/services/media_fields_catalog.py
"""Media fields visibility and categorization based on library type.

This module defines which fields are relevant for which library types,
enabling type-aware visibility control in statistics, search, and UI.
"""


# Fields that are relevant for each library type
FIELDS_BY_LIBRARY_TYPE = {
    "movies": {
        "video_fields": {
            "video_codec",
            "resolution",
            "hdr_type",
            "bitrate",
            "primary_video_width",
            "primary_video_height",
            "primary_video_resolution_pixels",
            "primary_video_hdr_type",
        },
        "audio_fields": {
            "audio_codec",
            "audio_spatial_profile",
            "audio_language",
            "audio_bitrate",
        },
        "shared_fields": {
            "container",
            "duration",
            "size",
            "quality_score",
            "file_path",
            "extension",
        },
    },
    "series": {
        "video_fields": {
            "video_codec",
            "resolution",
            "hdr_type",
            "bitrate",
            "primary_video_width",
            "primary_video_height",
            "primary_video_resolution_pixels",
            "primary_video_hdr_type",
        },
        "audio_fields": {
            "audio_codec",
            "audio_spatial_profile",
            "audio_language",
            "audio_bitrate",
        },
        "shared_fields": {
            "container",
            "duration",
            "size",
            "quality_score",
            "file_path",
            "extension",
        },
    },
    "music": {
        "video_fields": set(),
        "audio_fields": {
            "audio_codec",
            "audio_spatial_profile",
            "audio_language",
            "audio_bitrate",
            # Music-specific audio fields
            "audio_title",
            "audio_artist",
            "audio_album",
            "audio_album_artist",
            "audio_genre",
            "audio_date",
            "audio_disc",
            "audio_composer",
        },
        "shared_fields": {
            "container",
            "duration",
            "size",
            "quality_score",
            "file_path",
            "extension",
        },
    },
    "mixed": {
        "video_fields": {
            "video_codec",
            "resolution",
            "hdr_type",
            "bitrate",
            "primary_video_width",
            "primary_video_height",
            "primary_video_resolution_pixels",
            "primary_video_hdr_type",
        },
        "audio_fields": {
            "audio_codec",
            "audio_spatial_profile",
            "audio_language",
            "audio_bitrate",
            "audio_title",
            "audio_artist",
            "audio_album",
            "audio_album_artist",
            "audio_genre",
            "audio_date",
            "audio_disc",
            "audio_composer",
        },
        "shared_fields": {
            "container",
            "duration",
            "size",
            "quality_score",
            "file_path",
            "extension",
        },
    },
    "other": {
        "video_fields": {
            "video_codec",
            "resolution",
            "hdr_type",
            "bitrate",
            "primary_video_width",
            "primary_video_height",
            "primary_video_resolution_pixels",
            "primary_video_hdr_type",
        },
        "audio_fields": {
            "audio_codec",
            "audio_spatial_profile",
            "audio_language",
            "audio_bitrate",
            "audio_title",
            "audio_artist",
            "audio_album",
            "audio_album_artist",
            "audio_genre",
            "audio_date",
            "audio_disc",
            "audio_composer",
        },
        "shared_fields": {
            "container",
            "duration",
            "size",
            "quality_score",
            "file_path",
            "extension",
        },
    },
}


def get_visible_fields_for_library_type(library_type: str) -> set[str]:
    """Get all visible field identifiers for a library type.
    
    Args:
        library_type: One of "movies", "series", "music", "mixed", "other"
        
    Returns:
        Set of visible field identifiers
    """
    type_fields = FIELDS_BY_LIBRARY_TYPE.get(library_type, FIELDS_BY_LIBRARY_TYPE["other"])
    return (
        type_fields["shared_fields"]
        | type_fields["video_fields"]
        | type_fields["audio_fields"]
    )


def get_video_only_fields() -> set[str]:
    """Get fields that are video-only (hidden for music libraries)."""
    return {
        "video_codec",
        "resolution",
        "hdr_type",
        "bitrate",
        "primary_video_width",
        "primary_video_height",
        "primary_video_resolution_pixels",
        "primary_video_hdr_type",
    }


def get_music_only_fields() -> set[str]:
    """Get fields that are music-only (hidden for video-only libraries)."""
    return {
        "audio_title",
        "audio_artist",
        "audio_album",
        "audio_album_artist",
        "audio_genre",
        "audio_date",
        "audio_disc",
        "audio_composer",
    }


def should_hide_video_fields(library_type: str) -> bool:
    """Check if video-exclusive fields should be hidden for this library type."""
    return library_type == "music"


def should_hide_music_fields(library_type: str) -> bool:
    """Check if music-exclusive fields should be hidden for this library type."""
    return library_type in ("movies", "series")
